hooli-ai/hooli_ai/assets/ingestion.py

```diff
@@ -0,0 +1,481 @@
+    )
+
+    return github.convert_discussions_to_documents(discussions)
+
+
+@dg.asset(
+    group_name="embeddings",
+    kinds={"github", "openai", "pinecone"},
+    partitions_def=weekly_partition,
+    io_manager_key="document_io_manager",
+    automation_condition=dg.AutomationCondition.any_deps_updated(),
+    description="""
+   Creates vector embeddings from GitHub discussions and stores them in Pinecone.
+
+   This asset processes GitHub discussions by:
+   1. Converting discussion text into OpenAI embeddings
+   2. Storing embeddings and metadata in Pinecone vector database
+   3. Using namespace 'dagster-github' for discussions content
+
+   Dependencies:
+       - github_discussions_raw: Raw discussion documents to embed
+
+   Technical Details:
+       - Uses OpenAI's text-embedding-3-small model
+       - Embeddings dimension: 1536
+       - Stores in Pinecone index: 'dagster-knowledge'
+       - Includes metadata like title, URL, and creation date
+
+   Partitioning:
+       Uses weekly partitions to process discussions in batches
+
+   Returns:
+       MaterializeResult with metadata about number of discussions processed
+   """,
+)
+def github_discussions_embeddings(
+    context: dg.AssetExecutionContext,
+    openai: OpenAIResource,
+    pinecone: PineconeResource,
+    github_discussions_raw: list[Document],
+) -> dg.MaterializeResult:
+    # Create index if doesn't exist
+    pinecone.create_index("dagster-knowledge", dimension=1536)
+    index, namespace_kwargs = pinecone.get_index(
+        "dagster-knowledge", namespace="dagster-github"
+    )
+
+    splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
+        model_name=MODEL,
+        chunk_size=MAX_TOKENS,
+        chunk_overlap=0,
+    )
+    split_texts = []
+    for doc in github_discussions_raw:
+        split_texts.extend(splitter.split_text(doc.page_content))
+
+    all_embeddings = []
+    encoding = tiktoken.encoding_for_model(MODEL)
+
+    with openai.get_client(context) as client:
+        batch = []
+        token_count = 0
+
+        for text in split_texts:
+            text_tokens = len(encoding.encode(text))
+            if token_count + text_tokens > MAX_TOKENS:
+                # Send current batch
+                response = client.embeddings.create(
+                    model=MODEL,
+                    input=batch,
+                )
+                all_embeddings.extend([item.embedding for item in response.data])
+                time.sleep(1)
+                # Start new batch
+                batch = [text]
+                token_count = text_tokens
+            else:
+                batch.append(text)
+                token_count += text_tokens
+
+        # Final batch
+        if batch:
+            response = client.embeddings.create(
+                model=MODEL,
+                input=batch,
+            )
+            all_embeddings.extend([item.embedding for item in response.data])
```

Comment:
[1] was getting errors due to the discussion documents being too large so had to rework splitting and batching here